crates/trie/sparse/src/trie.rs

```diff
@@ -0,0 +1,615 @@
+        match node {
+            TrieNode::EmptyRoot => {
+                debug_assert!(path.is_empty());
+                self.nodes.insert(path, SparseNode::Empty);
+            }
+            TrieNode::Branch(branch) => {
+                let mut stack_ptr = branch.as_ref().first_child_index();
+                for idx in CHILD_INDEX_RANGE {
+                    if branch.state_mask.is_bit_set(idx) {
+                        let mut child_path = path.clone();
+                        child_path.push_unchecked(idx);
+                        self.reveal_node_or_hash(child_path, &branch.stack[stack_ptr])?;
+                        stack_ptr += 1;
+                    }
+                }
+                self.nodes
+                    .insert(path, SparseNode::Branch { state_mask: branch.state_mask, hash: None });
+            }
+            TrieNode::Extension(ext) => {
+                let mut child_path = path.clone();
+                child_path.extend_from_slice_unchecked(&ext.key);
+                self.reveal_node_or_hash(child_path, &ext.child)?;
+                self.nodes.insert(path, SparseNode::Extension { key: ext.key, hash: None });
+            }
+            TrieNode::Leaf(leaf) => {
+                let mut full = path.clone();
+                full.extend_from_slice_unchecked(&leaf.key);
+                self.values.insert(full, leaf.value);
+                self.nodes.insert(path, SparseNode::new_leaf(leaf.key));
+            }
+        }
+
+        Ok(())
+    }
+
+    fn reveal_node_or_hash(&mut self, path: Nibbles, child: &[u8]) -> alloy_rlp::Result<()> {
+        if child.len() == B256::len_bytes() + 1 {
+            // TODO: revise insert to not overwrite existing entries
+            self.nodes.insert(path, SparseNode::Hash(B256::from_slice(&child[1..])));
+            return Ok(())
+        }
+
+        self.reveal_node(path, TrieNode::decode(&mut &child[..])?)
+    }
+
+    /// Update the leaf node with provided value.
+    pub fn update_leaf(&mut self, path: Nibbles, value: Vec<u8>) {
+        self.prefix_set.insert(path.clone());
+        let existing = self.values.insert(path.clone(), value);
+        if existing.is_some() {
+            // trie structure unchanged, return immediately
+            return
+        }
+
+        let mut current = Nibbles::default();
+        while let Some(node) = self.nodes.get_mut(&current) {
+            match node {
+                SparseNode::Empty => {
+                    *node = SparseNode::new_leaf(path);
+                    break
+                }
+                SparseNode::Hash(_) => {
+                    unimplemented!() // TODO: error out
+                }
+                SparseNode::Leaf { key: current_key, .. } => {
+                    current.extend_from_slice_unchecked(current_key);
+
+                    // this leaf is being updated
+                    if current == path {
+                        // TODO: unreachable
```

Comment:
I don't quite understand why - is this because if a leaf node was already revealed and present at `path`, we would've just updated it above at line 159?

Reply:
correct